src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsService.kt

```diff
@@ -56,68 +64,178 @@ class LocationsService(
         saveMapping { mappingService.createMapping(it) }
       }
     } else {
-      telemetryClient.trackEvent("nonAssociation-create-ignored", telemetryMap)
+      telemetryClient.trackEvent("location-create-ignored", telemetryMap)
     }
   }
 
-  suspend fun deleteLocation(event: LocationDomainEvent) {
-    val dpsId = event.additionalInformation.id
-    val telemetryMap = mutableMapOf("dpsId" to dpsId)
+  suspend fun amendLocation(event: LocationDomainEvent) {
+    doUpdateLocation(event, "amend") { nomisLocationId, location ->
+      nomisApiService.updateLocation(nomisLocationId, toUpdateLocationRequest(location))
+    }
+  }
+
+  suspend fun deactivateLocation(event: LocationDomainEvent) {
+    doUpdateLocation(event, "deactivate") { nomisLocationId, location ->
+      nomisApiService.deactivateLocation(nomisLocationId, toDeactivateRequest(location))
+    }
+  }
+
+  suspend fun reactivateLocation(event: LocationDomainEvent) {
+    doUpdateLocation(event, "reactivate") { nomisLocationId, _ ->
+      nomisApiService.reactivateLocation(nomisLocationId)
+    }
+  }
+
+  suspend fun changeCapacity(event: LocationDomainEvent) {
+    doUpdateLocation(event, "capacity") { nomisLocationId, location ->
+      nomisApiService.updateLocationCapacity(nomisLocationId, UpdateCapacityRequest(location.capacity?.capacity, location.capacity?.operationalCapacity))
+    }
+  }
+
+  suspend fun changeCertification(event: LocationDomainEvent) {
+    doUpdateLocation(event, "certification") { nomisLocationId, location ->
+      nomisApiService.updateLocationCertification(
+        nomisLocationId,
+        UpdateCertificationRequest(location.certification?.capacityOfCertifiedCell ?: 0, location.certification?.certified ?: false),
+      )
+    }
+  }
+
+  suspend fun softDeleteLocation(event: LocationDomainEvent) {
+    val telemetryMap = mutableMapOf("dpsId" to event.additionalInformation.id)
+    telemetryMap["key"] = event.additionalInformation.key
     if (isDpsCreated(event.additionalInformation)) {
       runCatching {
-        val nomisId =
-          mappingService.getMappingGivenDpsId(dpsId).nomisLocationId
-            .also { telemetryMap += "nomisId" to it.toString() }
-
-        nomisApiService.deleteLocation(nomisId)
-        mappingService.deleteMapping(dpsId)
+        mappingService.getMappingGivenDpsId(event.additionalInformation.id).apply {
+          telemetryMap["nomisId"] = nomisLocationId.toString()
+          nomisApiService.deactivateLocation(nomisLocationId, DeactivateRequest())
```

Comment:
have you considered having a single gnomes api that do all this functions? Will you have any issues if some succeed and some fail or are they all idempotent?

Reply:
These are really just a placeholder as we are not yet sure how delete should be handled (except that 'real' delete is probably inadvisable due to references from other areas)